sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -760,7 +762,14 @@ private[hive] case class InsertIntoHiveTable(
   override def children: Seq[LogicalPlan] = child :: Nil
   override def output: Seq[Attribute] = child.output
 
-  override lazy val resolved: Boolean = childrenResolved && child.output.zip(table.output).forall {
+  val numDynamicPartitions = partition.values.count(_.isEmpty)
+  val tableOutput = {
+    val hiveTable = table.asInstanceOf[MetastoreRelation]
+    (hiveTable.attributes ++ hiveTable.partitionKeys.takeRight(numDynamicPartitions))
+      .take(child.output.length)
```

Comment:
Why a take here?  I haven't worked through it, but it seems possible that we are going to silently ignore the case where the number of columns doesn't line up.